File: algorithms/2_medium/M/03919_minimum_cost_to_move_between_indices/prefix_suffix.c
static inline int abs_int(const int a)
{
    return a < 0 ? -a : a;
}

/**
 * Note: The returned array must be malloced, assume caller calls free().
 */
int* minCost(const int* nums, const int len, int** queries, const int queries_len,
             const int* queries_col_len, int* return_len)
{
    assert(len >= 2 && queries_len > 0 && nums != NULL && queries != NULL && queries_col_len != NULL &&
           queries_col_len[0] == 2 && return_len != NULL);
    int* buff = calloc(2 * (size_t)len, sizeof(*buff));
    assert(buff != NULL);
    int* prefix = &buff[0];
    int* suffix = &buff[len];
    int* res = malloc((size_t)queries_len * sizeof(*res));
    assert(res != NULL);
    *return_len = queries_len;

    prefix[1] = 1;
    for (int i = 1; i < len - 1; i += 1) {
        assert(nums[i - 1] < nums[i] && nums[i] < nums[i + 1]);
        const int diff_l = abs_int(nums[i] - nums[i - 1]);
        const int diff_r = abs_int(nums[i + 1] - nums[i]);

        if (diff_l > diff_r) {
            prefix[i + 1] = 1 + prefix[i];
        } else {
            prefix[i + 1] = diff_r + prefix[i];
        }
    }

    suffix[len - 2] = 1;
    for (int i = len - 2; i >= 1; i -= 1) {
        const int diff_l = abs_int(nums[i] - nums[i - 1]);
        const int diff_r = abs_int(nums[i + 1] - nums[i]);

        if (diff_l <= diff_r) {
            suffix[i - 1] = 1 + suffix[i];
        } else {
            suffix[i - 1] = diff_l + suffix[i];
        }
    }

    for (int i = 0; i < queries_len; i += 1) {
        const int l = queries[i][0];
        const int r = queries[i][1];
        assert(l >= 0 && r >= 0 && l < len && r < len);

        if (l <= r) {
            res[i] = prefix[r] - prefix[l];
        } else {
            res[i] = suffix[r] - suffix[l];
        }
    }

/* cleanup: */
    free(buff);
    return res;
}

File: algorithms/2_medium/M/03919_minimum_cost_to_move_between_indices/prefix_suffix.rs
impl Solution {
    pub fn min_cost(nums: Vec<i32>, queries: Vec<Vec<i32>>) -> Vec<i32> {
        let len = nums.len();
        assert!(len >= 2);
        let mut buff = vec![0_i32; 2 * len];
        let (mut prefix, mut suffix) = buff.split_at_mut(len);
        let mut res = vec![0_i32; queries.len()];

        prefix[1] = 1;
        for i in 1..(len - 1) {
            debug_assert!(nums[i - 1] < nums[i] && nums[i] < nums[i + 1]);
            let diff_l = (nums[i] - nums[i - 1]).abs();
            let diff_r = (nums[i + 1] - nums[i]).abs();

            if diff_l > diff_r {
                prefix[i + 1] = 1 + prefix[i];
            } else {
                prefix[i + 1] = diff_r + prefix[i];
            }
        }

        suffix[len - 2] = 1;
        for i in (1..(len - 1)).rev() {
            let diff_l = (nums[i] - nums[i - 1]).abs();
            let diff_r = (nums[i + 1] - nums[i]).abs();

            if diff_l <= diff_r {
                suffix[i - 1] = 1 + suffix[i];
            } else {
                suffix[i - 1] = diff_l + suffix[i];
            }
        }

        for (i, q) in queries.into_iter().enumerate() {
            assert!(q.len() == 2);
            let (l, r) = (q[0], q[1]);
            assert!(l >= 0 && r >= 0 && (l as usize) < len && (r as usize) < len);

            if l <= r {
                res[i] = prefix[r as usize] - prefix[l as usize];
            } else {
                res[i] = suffix[r as usize] - suffix[l as usize];
            }
        }

        res
    }
}
